packages/react/modules/components/configure-adapter/configure-adapter.js

```diff
@@ -27,19 +23,96 @@ type Props = {
   defaultFlags?: Flags,
   children: React.Component<any>,
 };
+type State = {
+  adapterArgs: AdapterArgs,
+};
 type AdapterState = $Values<typeof AdapterStates>;
+type AdapterReconfigurationOptions = {
+  exact?: boolean,
+};
+type AdapterReconfiguration = {
+  adapterArgs: AdapterArgs,
+  options: AdapterReconfigurationOptions,
+};
+type ReconfigureAdapter = (
+  adapterArgs: AdapterArgs,
+  options: AdapterReconfigurationOptions
+) => void;
+
+export const AdapterContext: Context<ReconfigureAdapter> = createReactContext(
+  () => {}
+);
 
-export default class ConfigureAdapter extends PureComponent<Props> {
+export const mergeAdapterArgs = (
+  previousAdapterArgs: AdapterArgs,
+  { adapterArgs: nextAdapterArgs, options }: AdapterReconfiguration
+): AdapterArgs =>
+  options.exact ? nextAdapterArgs : merge(previousAdapterArgs, nextAdapterArgs);
+
+export default class ConfigureAdapter extends PureComponent<Props, State> {
   static defaultProps = {
     shouldDeferAdapterConfiguration: false,
     children: null,
     defaultFlags: {},
   };
 
   adapterState: AdapterState = AdapterStates.UNCONFIGURED;
-  setAdapterState = (nextAdapterState: AdapterState): void => {
+  pendingAdapterArgs: ?AdapterArgs = null;
+
+  state: { adapterArgs: AdapterArgs } = {
+    adapterArgs: this.props.adapterArgs,
+  };
+
+  setAdapterState = (
+    nextAdapterState: AdapterState,
+    exact: boolean = false
+  ): void => {
     this.adapterState = nextAdapterState;
   };
+  setAdapterArgs = (nextAdapterArgs: AdapterArgs): void =>
+    this.setState(prevState => ({
+      ...prevState,
+      adapterArgs: nextAdapterArgs,
+    }));
+
+  /**
+   * NOTE:
+   *   This is passed through the React context (it's a public API).
+   *   Internally this component has a `ReconfigureAdapter` type;
+   *   this function has two arguments for clarify.
+   */
+  reconfigureOrQueue = (
+    nextAdapterArgs: AdapterArgs,
+    options: AdapterReconfigurationOptions = {}
```

Comment:
Nitpick: you can have the default value in the `mergeAdapterArgs`, `reconfigureOrQueue` does not use `options`.